Sources/JNI/JNIObjects.swift

```diff
@@ -1,20 +1,29 @@
 import CJNI
+import Dispatch
 
 /// Designed to simplify calling a constructor and methods on a JavaClass
 /// Subclass this and add the methods appropriate to the object you are constructing.
 open class JNIObject {
-    public let javaClass: JavaClass
+    open class var className: String {
+        return "java.lang.object"
+    }
+
+    public class var javaClass: JavaClass {
+        return DispatchQueue.main.sync {
+            let javaClassLocalRef = try! jni.FindClass(name: className.replacingFullstopsWithSlashes())
+            try! checkAndThrowOnJNIError()
+            return jni.NewGlobalRef(javaClassLocalRef)!
+        }
+    }
+
     public let instance: JavaObject
 
     required public init(_ instance: JavaObject) throws {
         guard let globalInstanceRef = jni.NewGlobalRef(instance) else {
             throw Error.couldntCreateGlobalRef
         }
 
-        let javaClassLocalRef = try jni.GetObjectClass(obj: instance)
-        let javaClass = jni.NewGlobalRef(javaClassLocalRef)
         try checkAndThrowOnJNIError()
-        self.javaClass = javaClass!
         self.instance = globalInstanceRef
     }
 
```

Comment:
looks good overall, but we should deprecate the current convenience init and force use of the new API where users have to override `className` (it's weird to have a convenience init with `className` here)

Reply:
I think the best way to do that would be to add 

```
@available(*, deprecated, message: "Use the convenience initializer that doesn't take className as an argument")
convenience public init(_ className: String, arguments: [JavaParameterConvertible] = []) throws {
  // deprecated init
}

// note the variadic arguments syntax instead of the array
convenience public init(...arguments: [JavaParameterConvertible]) throws {
  // use `Self.javaClass` here
}
```

Reply:
i like the idea and i commited the changes. One thing, `Self` does not seem to be available here and i used `type(of: self).javaClass` to workaround it. Not really understanding why

Reply:
@ephemer